File: linker/lib/memory_map.rb
class MemoryMap
  
  def self.instance
    @@memory ||= MemoryMap.new
  end
  
  def self.clear!
    @@memory = MemoryMap.new
  end
  
  def self.create_program_module(base_address, symbols = {})
    program_module = ProgramModule.new(base_address)
    program_module.symbols = symbols
    MemoryMap.instance << program_module
  end
  
  def self.validate!
    uses = instance.modules.map { |m| m.uses }.flatten.uniq
    unused_definitions = SymbolTable.symbols.keys.find_all { |sym| !uses.include?(sym) }
    unused_definitions.sort.each do |sym|
      warnings << "Warning: #{sym} was defined in module #{definition_index(sym)} but never used."
    end
    instance.modules.each_with_index do |m, i|
      m.unused_symbols.each do |sym|
        warnings << "Warning: In module #{i + 1} #{sym} appeared in the use list but was not actually used."
      end
    end
  end
  
  def self.warnings
    instance.warnings
  end
  
  attr_reader :modules
  
  def initialize
    @modules = []
  end
  
  def [](index)
    @modules[index]
  end
  
  def map
    @modules.each { |m| m.map }
  end
  
  def to_s
    result = ["Memory Map"]
    @modules.each do |program_module|
      result << program_module.to_s
    end
    result = result.join("\n")
    result += "\n\n" + warnings.join("\n") if !warnings.empty?
    result
  end
  
  def warnings
    @warnings ||= []
  end
  
  def method_missing(sym, *args, &block)
    @modules.send sym, *args, &block
  end
  
protected
  
  def self.definition_index(symbol)
    module_index( instance.modules.detect { |pm| pm.defines?(symbol) } )
  end
  
  def self.module_index(program_module)
    instance.modules.index(program_module) + 1
  end
  
end

File: java/linker/src/lib/MemoryMap.java
package lib;

import java.util.Hashtable;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * User: Ross Kaffenberger
 * Date: Sep 22, 2009
 * Time: 11:35:43 PM
 */
public class MemoryMap {
    private static MemoryMap memoryMap = new MemoryMap();
    private ArrayList<ProgramModule> modules;
    private ArrayList<String> warnings;

    private MemoryMap() {
        modules = new ArrayList<ProgramModule>();
        warnings = new ArrayList<String>();
    }

    public static void validate() {
        ArrayList<String> uses = collectAllUses();

        for (String unusedDefinition : SymbolTable.symbols().keySet()) {
            if (!uses.contains(unusedDefinition)) {
                getWarnings().add("Warning: " + unusedDefinition + " was defined in module " + getModuleIndex(unusedDefinition)
                        + " but never used.");
            }
        }

        for (ProgramModule module : instanceModules()) {
            for(String unusedUse : module.unusedSymbols()) {
                getWarnings().add("Warning: In module "
                        + moduleIndex(module) + " "
                        + unusedUse + " appeared in the use list but was not actually used.");

            }
        }

    }

    private static ArrayList<String> collectAllUses() {
        ArrayList<String> uses = new ArrayList<String>();
        for (ProgramModule module : instanceModules()) {
            for (String use : module.getUses()) {
                if (!uses.contains(use)) {
                    uses.add(use);
                }
            }
        }
        return uses;
    }

    private static int getModuleIndex(String sym) {
        for (ProgramModule module : instanceModules()) {
            if (module.defines(sym)) {
                return moduleIndex(module);
            }
        }
        return 1;
    }

    private static int moduleIndex(ProgramModule module) {
        return instanceModules().indexOf(module) + 1;
    }

    private static ArrayList<ProgramModule> instanceModules() {
        return getInstance().getModules();
    }


    public static void createProgramModule(int baseAddress, Hashtable moduleSymbols) {
        ProgramModule module = new ProgramModule(baseAddress);
        module.setSymbols(moduleSymbols);
        MemoryMap.getInstance().add(module);
    }

    private void add(ProgramModule module) {
        this.modules.add(module);
    }

    public static MemoryMap getInstance() {
        return memoryMap;
    }

    public static ArrayList<String> getWarnings() {
        return getInstance().warnings;
    }

    public String toString() {
        String result = "Memory Map\n";
        for (ProgramModule module : modules) {
            result += module.toString() + "\n";
        }
        for (String warning : MemoryMap.getWarnings()) {
            result += "\n" + warning;
        }

        return result;
    }

    public ArrayList<ProgramModule> getModules() {
        return modules;
    }

    public static void clear() {
        memoryMap = new MemoryMap();
    }
}
